### PR TITLE
WIP [DO_NOT_MERGE]: Read registry configuration from Docker daemon in new-app

### PR DESCRIPTION
**Pull request by mfojtik**
This patch will use the engine-api client to get list of Docker registries (`ADD_REGISTRY`) configured for a Docker daemon and use that list to extended the search for the Docker image during the `new-app`.

In other words, if you have Dockerfile in your application source which has `FROM rhel7` and you have Docker that runs with `--add-registry registry.access.redhat.com` option, the `oc new-app` will include that registry when searching for image match.

@smarterclayton I think we should do something similar in image importer (we default to `docker.io` and not respecting the Docker configuration).

@csrwng you might have opinions about this.

Also fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1369769

### CODE REVIEW

**Comment by csrwng** [comment]
@mfojtik My concern would be that we're assuming that your local Docker is configured with the registries that you'd want to use in your cluster. It's highly likely that it is... just don't have a feel for a probability.

**Comment by mfojtik** [comment]
@csrwng we still search the cluster images first (if I understand that code right) and do this in a fallback. In that case, instead of just defaulting to "docker.io" which is hardcoded in importer, we at least try more registries. If there is a match, the IS created will point to that exact registry, so in case that registry does not exists in the cluster you will notice that, but we won't prevent your from creating the application because our resolution does not see registries other than "docker.io" when you don't specify the full path (which I think should be the "best practice" ;-)

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -202,6 +222,16 @@ func (c *AppConfig) SetOpenShiftClient(osclient client.Interface, OriginNamespac
 		ClientMapper: c.ClientMapper,
 		Namespace:    OriginNamespace,
 	}
+
+	// Attempt to discovert default registries
```
Comment (csrwng) [inline pkg/generate/app/cmd/newapp.go]: s/discovert/discover/

File: pkg/generate/app/dockerimagelookup.go [hunk 2]
```diff
@@ -196,7 +197,15 @@ func (s ImageImportSearcher) Search(precise bool, terms ...string) (ComponentMat
...
 		if err == client.ErrImageStreamImportUnsupported && s.Fallback != nil {
-			return s.Fallback.Search(precise, terms...)
+			extendedTerms := []string{}
+			for _, registryName := range s.Registries {
+				for _, imageName := range terms {
+					extendedTerms = append(extendedTerms, registryName+"/"+imageName)
+				}
+			}
+			extendedTerms = append(extendedTerms, terms...)
+			glog.V(5).Infof("fallbacking to generic search (%q, %#+v)", precise, extendedTerms)
```
Comment (csrwng) [inline pkg/generate/app/dockerimagelookup.go]: s/fallbacking/falling back/

**Comment by smarterclayton** [comment]
I find this highly unlikely.  Also, the image importer doesn't have
access to the docker socket, so it cannot do this.  If we aren't going
to do it on the server, doing it on the client doesn't help.

Why are we doing this again?

**Comment by mfojtik** [comment]
@smarterclayton https://bugzilla.redhat.com/show_bug.cgi?id=1369769 

basically you have Dockerfile with `FROM rhel7` and you run new-app locally and expect `rhel7` will be resolved properly if Docker runs with `-add-registry`

**Comment by smarterclayton** [comment]
First off - 90% of the systems out there will never have those aliases.
The remainder will be inconsistent.  Second, if the master needs to know
about this (and for the web console, it would have to), then you'd have to
add this on the master for image import.  Third, a much better workaround
is to create a "rhel" image stream in the "openshift" namespace - that
should work to satisfy the from.

Aliasing is a machine level decision, but should be a cluster level
decision.  Registry remapping is also somewhat of a code smell - if admins
want to fiddle with DNS names, we have other tools that can help with
this.  I'm neutral on aliasing on the master.

On Aug 25, 2016, at 10:15 AM, Michal Fojtik <notifications@github.com>
wrote:

@smarterclayton <https://github.com/smarterclayton>
https://bugzilla.redhat.com/show_bug.cgi?id=1369769

basically you have Dockerfile with FROM rhel7 and you run new-app locally
and expect rhel7 will be resolved properly if Docker runs with -add-registry

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/10637#issuecomment-242402081>,
or mute
the thread

**Comment by soltysh** [comment]
Based on previous comment, I'm closing this as this won't get fixed.
